### PR TITLE
Made many changes

### PR DESCRIPTION
**Pull request by daymondm**
Hi Tim, I've added many things to your existing code. I apologize in advance for some sloppy code (learning as I go) but everything as it stands has been reliably working. Modifications include:

- Ability to arm/disarm security system - the Vivint API handles system arming and disarming in a different manner than the rest of the devices. These states aren't tied to a specific device _id, but rather to the system itself. The modifications that I've made will add an accessory that is identified as a panel, and will inject a message to be handled by the panel _id. I'm not sure how this will behave if the user has multiple panels or if that's even possible. 
- All wireless sensors and locks now have a battery service that will show the battery level.
- Garage door opener functions as expected
- Each camera now adds a "PIV Detector" motion sensor. Every camera on the system that I've purchased has this flag set under the device snapshot, however I have noticed that only the indoor ping cameras will report a PIV true/false. The exterior cameras do not, or are terrible at detecting a person. The doorbell camera will report "vdt" instead of "piv", which are synonymous in this context. (person in view vs. visitor detected). The doorbell camera will not send a vdt: false, however. 

Thoughts for future features include as noted in your code: deleting unused accessories, and if the application loses connectivity it is unable to reconnect (e.g.: unplug router). Losing connectivity requires a sudo killall Homebridge to get things running again.

### CODE REVIEW

**Comment by mceres** [comment]
i updated my files with your code, but i am not seeing a panel device to arm disarm or battery levels.. i might be doing something wrong.

[4/13/2019, 2:51:05 PM] Loaded config.json with 0 accessories and 1 platforms.
[4/13/2019, 2:51:05 PM] ---
[4/13/2019, 2:51:06 PM] Loaded plugin: homebridge-vivint
[4/13/2019, 2:51:06 PM] Registering platform 'homebridge-vivint.Vivint'
[4/13/2019, 2:51:06 PM] ---
[4/13/2019, 2:51:06 PM] Loading 1 platforms...
[4/13/2019, 2:51:06 PM] [Vivint] Initializing Vivint platform...
[4/13/2019, 2:51:06 PM] Loading 0 accessories...
[4/13/2019, 2:51:07 PM] [Vivint] Adding 12 new accessories
[4/13/2019, 2:51:07 PM] [Vivint] { name: 'Master Bedroom Right Window',
  id: 23,
  deviceClassName: 'ContactSensor' }
[4/13/2019, 2:51:07 PM] [Vivint] { name: 'Master Bedroom Left Window',
  id: 25,
  deviceClassName: 'ContactSensor' }
[4/13/2019, 2:51:07 PM] [Vivint] { name: 'Master Bathroom Window One',
  id: 26,
  deviceClassName: 'ContactSensor' }
[4/13/2019, 2:51:07 PM] [Vivint] { name: 'Master Bathroom Two Window',
  id: 27,
  deviceClassName: 'ContactSensor' }
[4/13/2019, 2:51:07 PM] [Vivint] { name: 'Front Bedroom Window Right',
  id: 29,
  deviceClassName: 'ContactSensor' }
[4/13/2019, 2:51:07 PM] [Vivint] { name: 'Sliding Glass Door',
  id: 28,
  deviceClassName: 'ContactSensor' }
[4/13/2019, 2:51:07 PM] [Vivint] { name: 'Laundry Room Window',
  id: 31,
  deviceClassName: 'ContactSensor' }
[4/13/2019, 2:51:07 PM] [Vivint] { name: 'Front Bedroom Window Left',
  id: 30,
  deviceClassName: 'ContactSensor' }
[4/13/2019, 2:51:07 PM] [Vivint] { name: 'Front Door', id: 34, deviceClassName: 'ContactSensor' }
[4/13/2019, 2:51:07 PM] [Vivint] { name: 'Living Room (Living Room)',
  id: 39,
  deviceClassName: 'Thermostat' }
[4/13/2019, 2:51:07 PM] [Vivint] { name: 'Garage Lock', id: 72, deviceClassName: 'Lock' }
[4/13/2019, 2:51:07 PM] [Vivint] { name: 'Kids Bathroom Window',
  id: 83,
  deviceClassName: 'ContactSensor' }
HAP Warning: Characteristic D6D47D29-4642-4F44-B53C-D84015DAEBDB not in required or optional characteristics for service 0000004A-0000-1000-8000-0026BB765291. Adding anyway.
[4/13/2019, 2:51:07 PM] [Vivint] Handling incoming device snapshot for timestamp 2019-04-13T14:50:29.081000
status { category: 'PNConnectedCategory',
  operation: 'PNSubscribeOperation',
  affectedChannels: [ 'PlatformChannel#H4nf3iAXnxFdsSkptxuv51GZc7eLBpwQ' ],
  subscribedChannels: [ 'PlatformChannel#H4nf3iAXnxFdsSkptxuv51GZc7eLBpwQ' ],
  affectedChannelGroups: [],
  lastTimetoken: 0,
  currentTimetoken: '15551670291605235' }

**Comment by daymondm** [comment]
For the battery levels, you need to hard press on the contact sensor within HomeKit and click settings at the bottom. There should be a box that says battery level.. I haven’t figured out a way to make a battery level icon appear in HomeKit (if it’s even possible?). I’d need to see the devices that the plugin didn’t recognize to determine why it didn’t find your panel.. I’ll follow up later today

**Comment by mceres** [comment]
B1BDF02A-CB6F-47E0-BEFC-283AB469EBFA
FF8FAD21-AE88-49BF-9A42-6483D2B3BDAF

**Comment by mceres** [comment]
I checked there and no battery level and no panel to be able to arm and disarm :(

**Comment by mceres** [comment]
Correct me if I am wrong. But I just needed to replace the files that you updated inside the homebridge-vivint folder?

**Comment by mceres** [comment]
ok, i got it this time.. i was updating the files in the wrong path.. i figured it out.. i see the panel now, i can arm and disarm :)... also i can see the battery levels now.

here is the output that i get when loading it.. some of my sensors are not being detected, but i think it is just a matter of adding the device ids.

Great Job!

[4/14/2019, 8:09:37 PM] [Vivint] Do not know how to handle device! ID: 5 Data.t: energy_service data.ec: undefined Data.n: undefined
[4/14/2019, 8:09:37 PM] [Vivint] Do not know how to handle device! ID: 12 Data.t: network_hosts_service data.ec: undefined Data.n: undefined
[4/14/2019, 8:09:37 PM] [Vivint] Do not know how to handle device! ID: 13 Data.t: panel_diagnostics_service data.ec: undefined Data.n: undefined
[4/14/2019, 8:09:37 PM] [Vivint] Do not know how to handle device! ID: 32 Data.t: wireless_sensor data.ec: 655 Data.n: Back Bedroom Window
[4/14/2019, 8:09:37 PM] [Vivint] Do not know how to handle device! ID: 33 Data.t: wireless_sensor data.ec: 655 Data.n: Interior Garage Door
[4/14/2019, 8:09:37 PM] [Vivint] Do not know how to handle device! ID: 35 Data.t: wireless_sensor data.ec: 1248 Data.n: Bedroom Glass Break
[4/14/2019, 8:09:37 PM] [Vivint] Do not know how to handle device! ID: 36 Data.t: wireless_sensor data.ec: 1248 Data.n: Living Room Glass Break
[4/14/2019, 8:09:37 PM] [Vivint] Do not know how to handle device! ID: 38 Data.t: wireless_sensor data.ec: 609 Data.n: Living Room Motion Detector
[4/14/2019, 8:09:37 PM] [Vivint] Do not know how to handle device! ID: 44 Data.t: phillips_hue_bridge_device data.ec: undefined Data.n: Philips hue
[4/14/2019, 8:09:37 PM] [Vivint] Do not know how to handle device! ID: 45 Data.t: multilevel_switch_ip_device data.ec: undefined Data.n: Front Outdoor
[4/14/2019, 8:09:37 PM] [Vivint] Do not know how to handle device! ID: 46 Data.t: multilevel_switch_ip_device data.ec: undefined Data.n: Front Door Lights
[4/14/2019, 8:09:37 PM] [Vivint] Do not know how to handle device! ID: 47 Data.t: multilevel_switch_ip_device data.ec: undefined Data.n: Belen's
[4/14/2019, 8:09:37 PM] [Vivint] Do not know how to handle device! ID: 48 Data.t: multilevel_switch_ip_device data.ec: undefined Data.n: Cristian's
[4/14/2019, 8:09:37 PM] [Vivint] Do not know how to handle device! ID: 49 Data.t: multilevel_switch_ip_device data.ec: undefined Data.n: Master Bedroom
[4/14/2019, 8:09:37 PM] [Vivint] Do not know how to handle device! ID: 20003 Data.t: scheduler_service data.ec: undefined Data.n: undefined
[4/14/2019, 8:09:37 PM] [Vivint] Do not know how to handle device! ID: 70 Data.t: controller_device data.ec: undefined Data.n: Wink
[4/14/2019, 8:09:37 PM] [Vivint] Do not know how to handle device! ID: 75 Data.t: multilevel_switch_device data.ec: undefined Data.n: Living Room
[4/14/2019, 8:09:37 PM] [Vivint] Do not know how to handle device! ID: 79 Data.t: binary_switch_device data.ec: undefined Data.n: Garage
[4/14/2019, 8:09:37 PM] [Vivint] Do not know how to handle device! ID: 84 Data.t: wireless_sensor data.ec: 1269 Data.n: Master Hall Fire Alert
[4/14/2019, 8:09:37 PM] [Vivint] Do not know how to handle device! ID: 85 Data.t: multilevel_switch_device data.ec: undefined Data.n: Garage Entrance
[4/14/2019, 8:09:37 PM] [Vivint] Do not know how to handle device! ID: 89 Data.t: binary_switch_device data.ec: undefined Data.n: Play Area
[4/14/2019, 8:09:37 PM] [Vivint] Adding 14 new accessories
[4/14/2019, 8:09:37 PM] [Vivint] { name: 'Panel', id: 14, deviceClassName: 'Panel' }
[4/14/2019, 8:09:37 PM] [Vivint] { name: 'Master Bedroom Right Window',
  id: 23,
  deviceClassName: 'ContactSensor' }
[4/14/2019, 8:09:37 PM] [Vivint] { name: 'Master Bedroom Left Window',
  id: 25,
  deviceClassName: 'ContactSensor' }
[4/14/2019, 8:09:37 PM] [Vivint] { name: 'Master Bathroom Window One',
  id: 26,
  deviceClassName: 'ContactSensor' }
[4/14/2019, 8:09:37 PM] [Vivint] { name: 'Master Bathroom Two Window',
  id: 27,
  deviceClassName: 'ContactSensor' }
[4/14/2019, 8:09:37 PM] [Vivint] { name: 'Front Bedroom Window Right',
  id: 29,
  deviceClassName: 'ContactSensor' }
[4/14/2019, 8:09:37 PM] [Vivint] { name: 'Sliding Glass Door',
  id: 28,
  deviceClassName: 'ContactSensor' }
[4/14/2019, 8:09:37 PM] [Vivint] { name: 'Laundry Room Window',
  id: 31,
  deviceClassName: 'ContactSensor' }
[4/14/2019, 8:09:37 PM] [Vivint] { name: 'Front Bedroom Window Left',
  id: 30,
  deviceClassName: 'ContactSensor' }
[4/14/2019, 8:09:37 PM] [Vivint] { name: 'Front Door', id: 34, deviceClassName: 'ContactSensor' }
[4/14/2019, 8:09:37 PM] [Vivint] { name: 'Doorbell', id: 37, deviceClassName: 'PIVMotion' }
[4/14/2019, 8:09:37 PM] [Vivint] { name: 'Living Room (Living Room)',
  id: 39,
  deviceClassName: 'Thermostat' }
[4/14/2019, 8:09:37 PM] [Vivint] { name: 'Garage Lock', id: 72, deviceClassName: 'Lock' }
[4/14/2019, 8:09:37 PM] [Vivint] { name: 'Kids Bathroom Window',
  id: 83,
  deviceClassName: 'ContactSensor' }
HAP Warning: Characteristic D6D47D29-4642-4F44-B53C-D84015DAEBDB not in required or optional characteristics for service 0000004A-0000-1000-8000-0026BB765291. Adding anyway.
[4/14/2019, 8:09:37 PM] [Vivint] Handling incoming device snapshot for timestamp 2019-04-14T20:07:59.371000
status { category: 'PNConnectedCategory',
  operation: 'PNSubscribeOperation',
  affectedChannels: [ 'PlatformChannel#H4nf3iAXnxFdsSkptxuv51GZc7eLBpwQ' ],
  subscribedChannels: [ 'PlatformChannel#H4nf3iAXnxFdsSkptxuv51GZc7eLBpwQ' ],
  affectedChannelGroups: [],
  lastTimetoken: 0,
  currentTimetoken: '15552725777710563' }

**Comment by daymondm** [comment]
You could _probably_ change this line (229-232 device_set.js) to detect your back bedroom and and interior garage door contact sensors:

```
  extend(ContactSensor, {
    appliesTo: (data) => {
      return((data.t == "wireless_sensor") && ((data.ec == 1252) || (data.ec == 1251)))
    },
```

to:

```
  extend(ContactSensor, {
    appliesTo: (data) => {
      return((data.t == "wireless_sensor") && ((data.ec == 1252) || (data.ec == 1251) || (data.ec == 655))
    },
```

I'm presuming that they behave the same way as the rest of the contact sensors but unfortunately all of the sensors that I use are 1252/1251 sensors so I don't really have a way to test it out.

**Comment by mceres** [comment]
i was able to add them and test them, they are working fine. 

i had to arm and disarm the system to be able to see the changes of the sensors, for some reason it seems like if its going to sleep and not updating the status of the sensors. Any idea?

**Comment by daymondm** [comment]
You should be able to see all of the PubNub messages that has been sent/received by the panel in the Homebridge logfile. For debugging, I usually tail -f the log file and leave the terminal window open. I have noticed that sometimes the sensors are slow to update... especially if everything has been left alone for a while (i.e. in the mornings when I wake up, when I get home from work). I lowered the apiLoginRefreshSecs to 900 in my config.json file but I still experience the same. The sensor should eventually update if the PubNub message never gets sent or received when the api calls for a snapshot update. The snapshot contains status information for all of the Vivint sensors. 

I've found that relying on the Vivint sensors for automation works _most_ of the time, but not always. For example, I have an automation set to turn on the lights in the garage when I open the interior garage door. It seems to work if something has been recently controlled, but after I wake up and get ready for work, some times the lights turn on and some times it doesn't respond.

All that said, it seems like the sensors aren't extremely reliable for controlling automations. I'm not sure if its timing out or going to sleep like you suggested. Maybe adding controlling a device on a timer every 5-10 minutes or so will prevent it? I'm not sure; however, telling Siri to control things seems to work pretty reliably. (e.g.: opening the garage door, locking and unlocking doors)

**Comment by mceres** [comment]
Yes I was able to add them just fine. I will like to be able to add the zwave light switches. Going to look at the logs to see if I can figure something out or if you can point me in the right direction will be great. 

Also I got the following error and completely stop my homebridge. Not sure if it’s related to the changes you made or from original code. I can open a new issue if it’s better. 


29A44D22-DE48-487A-8F1C-16D7DEB011F4

**Comment by jlhunter2008** [comment]
@daymondm I added your code because I was wanting the arm/disarm feature. However, now my homebridge setup is in a constant rebooting loop. It does this about every 30 seconds. See error message below. Any direction in fixing this would be really helpful.
Screen Shot 2019-04-18 at 8 00 06 AM

**Comment by mceres** [comment]
It seems like maybe you have an instance of homebridge already running either on the background or from another terminal. That will be my guess. Maybe give it a reboot or check if there are other homebridge processes running and kill them.

**Comment by jlhunter2008** [comment]
@daymondm @mceres Deleting the cachedAccessories file resolved the issue. I have a question. What is the difference between "Night" and "Home" for the Panel?

**Comment by johnyoungberg** [comment]
I have been hoping that this would get merged into the master because I didn't want to mess with a working setup but it sounds like several people have had success manually moving to this setup.

What files do I need to replace to get on this version?

**Comment by daymondm** [comment]
@johnyoungberg device_set and vivint_api files are the only two changed. If you have a working set up you will need to delete cached accessories and reset it up for the battery levels to work properly

**Comment by daymondm** [comment]
@jlhunter2008 the night option works exactly the same as the home option. There is no difference, but HomeKit didn’t let me specify which arm states are available 

If armed at the panel, the code will always use home or away depending on which option you’ve chosen at the panel

**Comment by daymondm** [comment]
@mceres I haven’t yet experienced that error. Might try changing that line to this.renew()? I’ll have to try it out when I get back home. I think you received that error because the connection to the Vivint server was interrupted. WiFi went down for a bit or something like that.

**Comment by daymondm** [comment]
@mceres are the devices 45-49 your zwave switches? Do you get some sort of PubNub message with those Id#s in the log file when you turn the switch on and off? 

Actually if you could create an issue for it and hopefully I can be of some help in figuring it out. I enabled issues for my fork. I’m hoping it’ll be as simple as adding a line to appliesto similar to the contact sensors

**Comment by mceres** [comment]
@daymondm yes I do get a message. I created a new issue. Do you want to take a look there or do you want me to post it here?

Yes, they are zwave wall light switches. I have 2 that are just on and off, and 2 that are dimmers. 

I also have a glass break sensor and fire alarm if we want to keep on adding to this plug in and can do some testing.

**Comment by mceres** [comment]
@jlhunter2008 just in case you need to know the location of the files is /usr/local/lib/node_modules/homebridge-vivint

easiest way to do this is to copy them

vivint@homebridge-vivint:~$ sudo cp /source/device_set.js /usr/local/lib/node_modules/homebridge-vivint/lib/device_set.js
vivint@homebridge-vivint:~$ sudo cp /source/vivint_api.js /usr/local/lib/node_modules/homebridge-vivint/lib/vivint_api.js

File: lib/device_set.js [hunk 1]
```diff
@@ -37,30 +38,86 @@ function DeviceSetModule(config, log, homebridge, vivintApi, ThermostatCharacter
...
-          return;
-        }
-        msg.message.da.d.forEach((patch) => {
-          if (this.devicesById[patch._id]) {
-            this.devicesById[patch._id].handlePatch(patch)
-          }
-        })
-      }
+     
+     if (msg.message.da.seca) {
```
Comment (timcharper) [inline lib/device_set.js:81]: can you please extract this out to a function? We should keep this method simple.

Also, I wonder if we should process ARM / DISARM messages if the update is stale (we can receive messages that are older than the snapshot we achieved during initial auth and periodic re-auth)

File: lib/device_set.js [hunk 2]
```diff
@@ -142,8 +204,23 @@ function DeviceSetModule(config, log, homebridge, vivintApi, ThermostatCharacter
...
         return Characteristic.ContactSensorState.CONTACT_DETECTED
     }
+    
+    contactBatteryLevelValue() {
+	    return this.data.bl
+    }
 
     notify() {
+      if (this.contactBatteryLevelValue()) {
+	      if (0 <= this.contactBatteryLevelValue() && this.contactBatteryLevelValue() <= 10) {
```
Comment (timcharper) [inline lib/device_set.js:214]: Great! I'd suggest the 10 here should be a configurable value

File: lib/device_set.js [hunk 3]
```diff
@@ -282,16 +408,267 @@ function DeviceSetModule(config, log, homebridge, vivintApi, ThermostatCharacter
...
+		  * 1: ARM AWAY
+		  * 2: ARM NIGHT (Treated as ARM STAY)
+		  * 3: DISARM
+		  *
+		  * HomeKit apparently does not like it if you treat NIGHT_ARM the same as STAY_ARM
+		  *
+		  ****/
+
+
+      if (targetState === 3) {
```
Comment (timcharper) [inline lib/device_set.js:465]: perhaps we should use constants instead of numbers in code?

File: test1 [hunk 4]
```diff
@@ -0,0 +1 @@
+test1
```
Comment (timcharper) [inline test1:1]: did you mean to add this?

File: device_set.js [hunk 5]
```diff
@@ -0,0 +1,923 @@
+const dataPatch = require("../lib/datapatch.js")
```
Comment (timcharper) [inline device_set.js:1]: Why are there two copies of this file?

**Review by timcharper — CHANGES_REQUESTED** [review]

**Comment by timcharper** [comment]
Hey @daymondm ! I just noticed this, thanks for the PR! I left a handful of comments. I don't think this is quite ready to merge but it looks really good!

**Comment by daymondm** [comment]
Thanks @timcharper for the feedback. Ill submit a new pull request when completed!

**Pull request closed without merging**